src/webgl/3d_primitives.js

```diff
@@ -10,6 +10,179 @@ import p5 from '../core/main';
 
+/**
+ * Starts creating a new p5.Geometry. Subsequent shapes drawn will be added
+ * to the geometry and then returned when
+ * <a href="#/p5/endGeometry">endGeometry()</a> is called. One can also use
+ * <a href="#/p5/buildGeometry">buildGeometry()</a> to pass a function that
+ * draws shapes.
+ *
+ * @method beginGeometry
+ *
+ * @example
+ * <div>
+ * <code>
+ * let shapes;
+ *
+ * function setup() {
+ *   createCanvas(100, 100, WEBGL);
+ *   makeShapes();
+ * }
+ *
+ * function makeShapes() {
+ *   beginGeometry();
+ *   scale(0.18);
+ *
+ *   push();
+ *   translate(100, -50);
+ *   scale(0.5);
+ *   rotateX(PI/4);
+ *   cone();
+ *   pop();
+ *   cone();
+ *
+ *   beginShape();
+ *   vertex(-20, -50);
+ *   quadraticVertex(
+ *     -40, -70,
+ *     0, -60
+ *   );
+ *   endShape();
+ *
+ *   beginShape(TRIANGLE_STRIP);
+ *   for (let y = 20; y <= 60; y += 10) {
+ *     for (let x of [20, 60]) {
+ *       vertex(x, y);
+ *     }
+ *   }
+ *   endShape();
+ *
+ *   beginShape();
+ *   vertex(-100, -120);
+ *   vertex(-120, -110);
+ *   vertex(-105, -100);
+ *   endShape();
+ *
+ *   shapes = endGeometry();
+ * }
+ *
+ * function draw() {
+ *   background(255);
+ *   lights();
+ *   orbitControl();
+ *   model(shapes);
+ * }
+ * </code>
+ * </div>
+ *
+ * @alt
+ * A series of different flat, curved, and 3D shapes floating in space.
+ */
+p5.prototype.beginGeometry = function() {
+  return this._renderer.beginGeometry();
+};
+
+/**
+ * Finishes creating a new <a href="#/p5.Geometry">p5.Geometry</a> that was
+ * started using <a href="#/p5/beginGeometry">beginGeometry()</a>. One can also
+ * use <a href="#/p5/buildGeometry">buildGeometry()</a> to pass a function that
+ * draws shapes.
+ *
+ * @method endGeometry
+ * @returns {p5.Geometry} The model that was built.
+ */
+p5.prototype.endGeometry = function() {
+  return this._renderer.endGeometry();
+};
+
+/**
+ * Creates a new <a href="#/p5.Geometry">p5.Geometry</a> that contains all
+ * the shapes drawn in a provided callback function. The returned combined shape
+ * can then be drawn all at once using <a href="#/p5/model">model()</a>.
+ *
+ * If you need to draw complex shapes every frame which don't change over time,
+ * combining them with `buildGeometry()` once and then drawing that will likely
+ * run faster than repeatedly drawing the individual pieces.
+ *
+ * One can also draw shapes directly between
+ * <a href="#/p5/beginGeometry">beginGeometry()</a> and
+ * <a href="#/p5/endGeometry">endGeometry()</a> instead of using a callback
+ * function.
+ *
+ * @method buildGeometry
+ * @param {Function} callback A function that draws shapes.
+ * @returns {p5.Geometry} The model that was built from the callback function.
+ *
+ * @example
+ * <div>
+ * <code>
+ * let tree;
+ * let button;
+ *
+ * function setup() {
+ *   createCanvas(100, 100, WEBGL);
+ *   button = createButton('New');
+ *   button.mousePressed(makeTree);
+ *   makeTree();
+ * }
+ *
+ * function makeTree() {
```

Comment:
It's a cool example, but it requires a user to know about recursion. Is there a simpler way to demonstrate the feature?

Reply:
I'm going to replace it with a bunch of spheres in a Gaussian distribution cloud, just using a loop instead of recursion: https://editor.p5js.org/davepagurek/sketches/bxXK1bLVp